p2panda-rs/src/storage_provider/traits/storage_provider.rs

```diff
@@ -0,0 +1,494 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+
+use async_trait::async_trait;
+
+use crate::document::DocumentId;
+use crate::entry::{decode_entry, SeqNum};
+use crate::hash::Hash;
+use crate::operation::{AsOperation, Operation};
+use crate::storage_provider::errors::PublishEntryError;
+use crate::storage_provider::models::{EntryWithOperation, Log};
+use crate::storage_provider::traits::{
+    AsEntryArgsRequest, AsEntryArgsResponse, AsPublishEntryRequest, AsPublishEntryResponse,
+    AsStorageEntry, AsStorageLog, EntryStore, LogStore,
+};
+
+/// Trait which handles all high level storage queries and insertions.
+///
+/// This trait should be implemented on the root storage provider struct. It's definitions make up
+/// the the higher level methods a p2panda client needs for interacting with data storage.
+#[async_trait]
+pub trait StorageProvider<StorageEntry: AsStorageEntry, StorageLog: AsStorageLog>:
+    EntryStore<StorageEntry> + LogStore<StorageLog>
+{
+    /// Params when making a request to `get_entry_args`.
+    type EntryArgsRequest: AsEntryArgsRequest + Sync;
+
+    /// Response from a call to `get_entry_args`.
+    type EntryArgsResponse: AsEntryArgsResponse;
+
+    /// Params when making a request to `publish_entry`.
+    type PublishEntryRequest: AsPublishEntryRequest + Sync;
+
+    /// Response from a call to `publish_entry`.
+    type PublishEntryResponse: AsPublishEntryResponse;
+
+    /// Returns the related document for any entry.
+    ///
+    /// Every entry is part of a document and, through that, associated with a specific log id used
+    /// by this document and author. This method returns that document id by looking up the log
+    /// that the entry was stored in.
+    async fn get_document_by_entry(
+        &self,
+        entry_hash: &Hash,
+    ) -> Result<Option<DocumentId>, Box<dyn std::error::Error>>;
```

Comment:
Rust noob question: what is this strange error type and why doesn't this use an error type defined in `errors.rs` like the other structs of the module?

Reply:
Could be helpful to add a note to the docstring and describe under what conditions it would be okay to return `None`

Reply:
This is a lovely boxed error, which means it can be anything ;-p box is a smart pointer thingy.

I'm not sure it's the right pattern here actually though, i wanted to allow for implementors to be able to return their own errors from these higher level methods. I think we could use a similar "error wrapping" technique as in the lower level traits as well, but this seemed to work a little better for me.

Reply:
But the docstring should explain the valid conditions for a `None` return value don't you think? At least it's not obvious for me from this alone.